### PR TITLE
Process career quiz results by returning most career path most frequently associated with user choices

### PR DESCRIPTION
**Pull request by lzoebaker**

### CODE REVIEW

File: src/main/java/com/google/sps/data/ProcessCareerQuizResults.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+        choiceFrequencyHashMap.put(careerPath, choiceFrequencyHashMap.get(careerPath) + 1); 
+      } 
+      else { 
+        choiceFrequencyHashMap.put(careerPath, 1);
+      } 
+    }
+    return mostFrequentInHashMap(choiceFrequencyHashMap);
+  }
+
+  private static String mostFrequentInHashMap(Map<String, Integer> hashMap) {
```
Comment (WillHipschman) [inline src/main/java/com/google/sps/data/ProcessCareerQuizResults.java:30]: name this "choiceFrequencyMap"

Comment (WillHipschman) [inline src/main/java/com/google/sps/data/ProcessCareerQuizResults.java:30]: Name the method "getMostFrequentChoice" or a name that more clearly indicates what it does.

File: src/main/java/com/google/sps/data/ProcessCareerQuizResults.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+import java.util.HashMap; 
+import java.util.Map; 
+import java.util.Map.Entry;
+import com.google.sps.data.CareerQuestionChoice;
+
+/** Returns the recommended career path based off the career quiz results. */
+public class ProcessCareerQuizResults {
+  private ProcessCareerQuizResults() {};
+  
+  public static String getRecommendedCareerPath(List<CareerQuestionChoice> userChoices) {
```
Comment (WillHipschman) [inline src/main/java/com/google/sps/data/ProcessCareerQuizResults.java:13]: This looks a lot like my interview question :)

File: src/test/java/com/google/sps/RecommendedCareerPathTest.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+import org.junit.runners.JUnit4;
+import com.google.sps.data.CareerQuestionChoice;
+import com.google.sps.data.CareerQuestionAndChoices;
+import java.util.List;
+import java.util.Arrays;
+
+
+/** Tests the CareerQuiz servlet and its interactions with CareerQuestionDatabase*/
+@RunWith(JUnit4.class)
+public final class RecommendedCareerPathTest {
```
Comment (WillHipschman) [inline src/test/java/com/google/sps/RecommendedCareerPathTest.java:16]: Include a test for an empty input list

File: src/main/java/com/google/sps/data/ProcessCareerQuizResults.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    Map<String, Integer> choiceFrequencyHashMap = new HashMap<String, Integer>(); 
+    String careerPath;
+    String maxFrequencyCareerPath = new String();
+    // creates a frequency hashmap mapping associated career paths of choices to their frequencies
+    for (CareerQuestionChoice choice : userChoices) {
+      careerPath = choice.getAssociatedCareerPath();
+      if (choiceFrequencyHashMap.containsKey(careerPath)) {
+        choiceFrequencyHashMap.put(careerPath, choiceFrequencyHashMap.get(careerPath) + 1); 
+      } 
+      else {
```
Comment (DirtGrubDylan) [inline src/main/java/com/google/sps/data/ProcessCareerQuizResults.java:23]: `} else {`

File: src/main/java/com/google/sps/data/ProcessCareerQuizResults.java [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+    String maxFrequencyCareerPath = new String();
+    // creates a frequency hashmap mapping associated career paths of choices to their frequencies
+    for (CareerQuestionChoice choice : userChoices) {
+      careerPath = choice.getAssociatedCareerPath();
+      if (choiceFrequencyHashMap.containsKey(careerPath)) {
+        choiceFrequencyHashMap.put(careerPath, choiceFrequencyHashMap.get(careerPath) + 1); 
+      } 
+      else { 
+        choiceFrequencyHashMap.put(careerPath, 1);
+      }
```
Comment (DirtGrubDylan) [inline src/main/java/com/google/sps/data/ProcessCareerQuizResults.java:25]: This `if/else` block can be reduced with this handy little function:
```
		Integer count = map.getOrDefault(key, 0);
		map.put(key, count + 1);
```

File: src/main/java/com/google/sps/data/ProcessCareerQuizResults.java [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+    String mostFrequentKey = "";
+    for (String key : hashMap.keySet()) {
+      currentFrequency = hashMap.get(key);
+      if (currentFrequency > maxFrequencyCount) {
+        mostFrequentKey = key;
+        maxFrequencyCount = currentFrequency;
+      }
+    }
+    return mostFrequentKey;
+  }
```
Comment (DirtGrubDylan) [inline src/main/java/com/google/sps/data/ProcessCareerQuizResults.java:42]: This can be reduced to something like so:
```
    Entry<K, V> maxEntry = Collections.max(map.entrySet(), (Entry<K, V> e1, Entry<K, V> e2) -> e1.getValue()
        .compareTo(e2.getValue()));
    return maxEntry.getValue();
```

You can choose to do this, or keep what you have. It's more about clarity at this point.

File: src/test/java/com/google/sps/RecommendedCareerPathTest.java (hunk 3 above)
Comment (DirtGrubDylan) [reply src/test/java/com/google/sps/RecommendedCareerPathTest.java:16]: +1

File: src/test/java/com/google/sps/RecommendedCareerPathTest.java [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+  @Test
+  public void outputsMostFrequentCareerPath_2choices() {
+    List<CareerQuestionChoice> userChoices = Arrays.asList(
+      new CareerQuestionChoice(CHOICE_1, CAREER_1),
+      new CareerQuestionChoice(CHOICE_2, CAREER_2),
+      new CareerQuestionChoice(CHOICE_3, CAREER_2),
+      new CareerQuestionChoice(CHOICE_4, CAREER_3)
+    );
+    String recommendedCareerPath = ProcessCareerQuizResults.getRecommendedCareerPath(userChoices);
+    Assert.assertEquals(recommendedCareerPath, CAREER_2);
```
Comment (DirtGrubDylan) [inline src/test/java/com/google/sps/RecommendedCareerPathTest.java:36]: So I would structure tests to be more like:
```
  @Test
  public void outputsMostFrequentCareerPath_2choices() {
    List<CareerQuestionChoice> userChoices = Arrays.asList(
      new CareerQuestionChoice(CHOICE_1, CAREER_1),
      new CareerQuestionChoice(CHOICE_2, CAREER_2),
      new CareerQuestionChoice(CHOICE_3, CAREER_2),
      new CareerQuestionChoice(CHOICE_4, CAREER_3)
    );

    String expected = CAREER_2;

    String result = ProcessCareerQuizResults.getRecommendedCareerPath(userChoices);

    Assert.assertEquals(result, expected);
  }
```

Comment (DirtGrubDylan) [reply src/test/java/com/google/sps/RecommendedCareerPathTest.java:36]: You can leave it as is if you feel more comfortable.

File: src/test/java/com/google/sps/RecommendedCareerPathTest.java [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+      new CareerQuestionChoice(CHOICE_3, CAREER_2),
+      new CareerQuestionChoice(CHOICE_4, CAREER_2),
+      new CareerQuestionChoice(CHOICE_5, CAREER_3)
+    );
+    String recommendedCareerPath = ProcessCareerQuizResults.getRecommendedCareerPath(userChoices);
+    Boolean oneOfTopIsRecommended = false;
+    if (recommendedCareerPath.equals(CAREER_1) || recommendedCareerPath.equals(CAREER_2)) {
+        oneOfTopIsRecommended = true;
+    }
+    Assert.assertTrue(oneOfTopIsRecommended);
```
Comment (DirtGrubDylan) [inline src/test/java/com/google/sps/RecommendedCareerPathTest.java:53]:
```
Assert.assertTrue(recommendedCareerPath.equals(CAREER_1) || recommendedCareerPath.equals(CAREER_2));
```

File: src/main/java/com/google/sps/data/ProcessCareerQuizResults.java (hunk 1 above)
Comment (lzoebaker) [reply src/main/java/com/google/sps/data/ProcessCareerQuizResults.java:30]: done.

Comment (lzoebaker) [reply src/main/java/com/google/sps/data/ProcessCareerQuizResults.java:30]: done.

File: src/main/java/com/google/sps/data/ProcessCareerQuizResults.java (hunk 2 above)
Comment (lzoebaker) [reply src/main/java/com/google/sps/data/ProcessCareerQuizResults.java:13]: ;)

File: src/main/java/com/google/sps/data/ProcessCareerQuizResults.java (hunk 4 above)
Comment (lzoebaker) [reply src/main/java/com/google/sps/data/ProcessCareerQuizResults.java:23]: done.

File: src/main/java/com/google/sps/data/ProcessCareerQuizResults.java (hunk 5 above)
Comment (lzoebaker) [reply src/main/java/com/google/sps/data/ProcessCareerQuizResults.java:25]: very nifty, done.

File: src/main/java/com/google/sps/data/ProcessCareerQuizResults.java (hunk 6 above)
Comment (lzoebaker) [reply src/main/java/com/google/sps/data/ProcessCareerQuizResults.java:42]: Very cool! I might keep what I have just for clarity.

File: src/test/java/com/google/sps/RecommendedCareerPathTest.java (hunk 7 above)
Comment (lzoebaker) [reply src/test/java/com/google/sps/RecommendedCareerPathTest.java:36]: I like the clarity of this structure, changed.

File: src/test/java/com/google/sps/RecommendedCareerPathTest.java (hunk 8 above)
Comment (lzoebaker) [reply src/test/java/com/google/sps/RecommendedCareerPathTest.java:53]: done.

File: src/test/java/com/google/sps/RecommendedCareerPathTest.java (hunk 3 above)
Comment (lzoebaker) [reply src/test/java/com/google/sps/RecommendedCareerPathTest.java:16]: added.
